### ISSUE TITLE
consider relaxing requirement that '@timestamp', 'log.level', and 'message' are always first

### ISSUE THREAD

**Issue by trentm** [elastic/ecs-logging#42]
https://www.elastic.co/guide/en/ecs-logging/overview/master/intro.html#_why_ecs_logging says:

>  Decently human-readable JSON structure
>
>    The first three fields are always @timestamp, log.level and message. ...

The following logging frameworks do not currently support controlling the order of log record fields, so cannot reasonably support this as a strong requirement:

- Go's zap (https://github.com/elastic/ecs-logging-go-zap)
- Node.js's pino (https://github.com/pinojs/pino, https://github.com/elastic/ecs-logging-js/tree/master/loggers/pino)

This issue is to collect notes and concerns with relaxing the spec and language around these fields always having to be first.

**Comment by trentm** [comment]
## regarding pino

(From a chat discussion.)

> I’d be curious why the order in [...] pino can’t be controlled.

tl;dr: Pino just doesn't expose controlling the order of log record fields.

Details: Pino's raison d'etre as a logging framework is to be as fast as possible. That means it does as little JSON re-serialization of fields as possible. It doesn't build up a mapping of fields to emit for each log record and then JSON serialize that. Instead it breaks a log record into (hopefully) independent chunks:

```
<level>,<time>,<other fields>,<message>
```

Because "speed" it doesn't ensure there are no collisions of keys in those "<other fields>", so I can get a collision of the "time" field if I do this:

```
logger.info({time: 'foo'}, 'hi again')
```

resulting in:

```
{"level":30,"time":1610387237919,"pid":44398,"hostname":"pink.local","time":"foo","msg":"hi again"}
```

While I'm not sure it was intentional, putting "msg" *last* ensures that a typical JSON parser will make that last one "win" if there are duplicate keys.

From Pino's docs: https://getpino.io/#/docs/child-loggers?id=duplicate-keys-caveat

> Ultimately the conflict is resolved by taking the last value ...
> There may be cases where this edge case becomes problematic if a JSON parser with alternative behavior is used to process the logs.

We are hoping that ES ingest is fine with cases like this.

**Comment by felixbarny** [comment]
When I wrote the spec, I didn't know that there were logging libraries that don't allow to modify the order of fields.

The reason to specify the order is that the most important fields are shown upfront so that the logs can be read in the terminal by just using `tail` or `less`.

I'd like to keep the order in the spec but loggers that don't support to specify the order of the fields can ignore that part of the spec. So it's something in-between MUST and SHOULD. For loggers that do support field ordering, it would be a MUST. For loggers that don't support field ordering, this part of the spec just doesn't apply.

Do you have a suggestion on how to improve the wording on the docs to reflect this?

Another thing we might want to consider in the future is creating a (go-based?) CLI tool that helps with making ecs-json logs even more human-readable, similar to https://github.com/trentm/node-bunyan. Ideally, that tool would be able to be used in combination with `less` and `tail`.
But even when having that tool I think there's value in making the logs decently human-readable when that tool is not installed.

**Comment by trentm** [comment]
> So it's something in-between MUST and SHOULD

Heh. "SHOULD IN THE STRONGEST POSSIBLE TERMS".



> Another thing we might want to consider in the future is creating a (go-based?) CLI tool 

Yes definitely. I'm ... keen, but I've stayed away from it right now because I'd be lost in the rabbit hole for a while.
I think I could fairly quickly adapt a version of `bunyan` to work with ecs-logging format. However, I agree that something implemented in node.js wouldn't have as wide an appeal. If implemented it Go (or Rust, say), I'd be a slower starter on it.

> Do you have a suggestion on how to improve the wording on the docs to reflect this?

I'll have a go at a PR.
